File: app/metrics/info.py
"""Info line of a link tile: two or three short facts per service.

    parts = [{"key": "kimai.today", "params": {"hours": "3,5"}}]   (key → catalog "info.<key>")
"""

from datetime import UTC, date, datetime

from app.metrics import kimai as km
from app.metrics import ninja as nm
from app.metrics import snipe as sm
from app.metrics.dates import parse_time

MINUTES_PER_HOUR = 60


def _part(key: str, **params) -> dict:
    return {"key": key, "params": params}


def parts(service: str, data: dict, today: date) -> list[dict]:
    builder = {"kimai": _kimai, "invoiceninja": _ninja, "snipeit": _snipe, "dawarich": _dawarich,
               "glances": _glances}.get(service)
    return builder(data, today) if builder else []


def _kimai(data: dict, today: date) -> list[dict]:
    stats = km.summary(data, today)
    found = [_part("kimai.today", hours={"$num": stats["today_min"] / MINUTES_PER_HOUR, "digits": 1})]
    if stats["running"]:
        found.append(_part("kimai.running"))
    return found


def _ninja(data: dict, today: date) -> list[dict]:
    stats = nm.summary(data, today)
    found = []
    if stats["overdue"]:
        found.append(_part("in.overdue", count=len(stats["overdue"])))
    found.append(_part("in.open", amount={"$money": stats["open_amount"], "currency": stats["currency"]}))
    return found


def _snipe(data: dict, today: date) -> list[dict]:
    stats = sm.summary(data, today)
    return [_part("snipe.assets", count=stats["assets"]),
            _part("snipe.upcoming", count=len(stats["upcoming"]))]


def _dawarich(data: dict, today: date) -> list[dict]:
    last = parse_time(data.get("last_point"))
    if last is None:
        return []
    if last.tzinfo is None:
        last = last.replace(tzinfo=UTC)
    hours = int((datetime.now(UTC) - last).total_seconds() // 3600)
    return [_part("dawarich.last", hours=hours)]


def _glances(data: dict, today: date) -> list[dict]:
    return [_part("glances.cpu", percent=round(data.get("cpu") or 0))]

File: internal/metrics/info.go
// Package metrics: the info line of a link tile — two or three short facts
// per service. Params use the same typed shape as hint params
// ({"$num": ...}, {"$money": ...}), formatted per reader locale by i18n.Typed.
package metrics

import (
	"time"

	"andon/internal/sources"
)

// InfoPart is one fact: key -> catalog "info.<key>", with typed params.
type InfoPart struct {
	Key    string
	Params map[string]any
}

func part(key string, params map[string]any) InfoPart {
	if params == nil {
		params = map[string]any{}
	}
	return InfoPart{Key: key, Params: params}
}

// KimaiInfo builds the info line for a Kimai link tile.
func KimaiInfo(data *sources.KimaiDataset, today time.Time) []InfoPart {
	stats := KimaiSummaryOf(data, today, 0)
	found := []InfoPart{part("kimai.today", map[string]any{
		"hours": map[string]any{"$num": float64(stats.TodayMin) / minutesPerHour, "digits": 1},
	})}
	if len(stats.Running) > 0 {
		found = append(found, part("kimai.running", nil))
	}
	return found
}

// NinjaInfo builds the info line for an Invoice Ninja link tile.
func NinjaInfo(data *sources.NinjaDataset, today time.Time) []InfoPart {
	stats := NinjaSummaryOf(data, today, "", "")
	var found []InfoPart
	if len(stats.Overdue) > 0 {
		found = append(found, part("in.overdue", map[string]any{"count": len(stats.Overdue)}))
	}
	found = append(found, part("in.open", map[string]any{
		"amount": map[string]any{"$money": stats.OpenAmount, "currency": stats.Currency},
	}))
	return found
}

// SnipeInfo builds the info line for a Snipe-IT link tile.
func SnipeInfo(data *sources.SnipeDataset, today time.Time) []InfoPart {
	stats := SnipeSummaryOf(data, today)
	return []InfoPart{
		part("snipe.assets", map[string]any{"count": stats.Assets}),
		part("snipe.upcoming", map[string]any{"count": len(stats.Upcoming)}),
	}
}

// DawarichInfo builds the info line for a Dawarich link tile.
func DawarichInfo(lastPoint string) []InfoPart {
	last, ok := ParseTime(lastPoint)
	if !ok {
		return nil
	}
	hours := int(time.Now().UTC().Sub(last).Hours())
	return []InfoPart{part("dawarich.last", map[string]any{"hours": hours})}
}

// GlancesInfo builds the info line for a Glances link tile.
func GlancesInfo(cpuPercent float64) []InfoPart {
	return []InfoPart{part("glances.cpu", map[string]any{"percent": round(cpuPercent)})}
}
